### ISSUE TITLE
Trains fly for longer than expected

### ISSUE THREAD

**Issue by Xcelled** [Kiplacon/RenaiTransportation#16]
(Continuing from https://github.com/robot256/MultipleUnitTrainControl/issues/14 because the bug appears to be in RenaiTransportation)

I added some code to warn when trains were still in the air after their air time had elapsed. This apparently happens a lot, and I wouldn't be surprised if nuclear fuel made it worse to the point that the train actually breaks apart upon landing due to the speed differentials:

![image](https://user-images.githubusercontent.com/5599341/99591579-02f37700-29bd-11eb-94ed-36fe699da721.png)

**Comment by Kiplacon** [comment]
How did you determine how much longer than expected the train took to land?

**Comment by Xcelled** [comment]
I stuck this in the tick handler for trains:

```lua
		if (game.tick - properties.LaunchTick) > properties.AirTime then
			game.print("WARNING: Train " .. PropUnitNumber .. " in the air for " .. (game.tick - properties.LaunchTick) - properties.AirTime .. " ticks longer than expected")
		end
```

but I just realized I stuck it outside the landing check if, and trains aren't removed until the entire train has landed :flushed: The message still gets printed a lot, but with only one tick of difference.

**Comment by Xcelled** [comment]
One thing I noticed about trains that break p is that they actually break up immediately upon launching, not upon landing. They also "count" as multiple entities as far as magnet ramps are concerned so they trigger the notice about depleted buffers:

![image](https://user-images.githubusercontent.com/5599341/99602607-c41aec80-29cf-11eb-9e4a-b71dcfe9d18c.png)

**Comment by Kiplacon** [comment]
There are two things I can think of that might be related to this.

1. When calculating when the train is going to land, the result of the calculation is rounded up to the nearest whole number since ticks are whole numbers. Since it's just rounding up it should only add on an extra tick, not 2-20 like in your screenshot, but maybe there is something else amplifying that rounding.

2. When a train hits the ramp, it doesn't actually maintain full speed. It slows down ever so slightly because it just hit something and that slightly slower speed is what is used to calculate the landing tick, pull the following carriages, and push the landing carriages. Since landing tick is calculated based on the train's speed, that slight reduction in speed on impact might make more of a difference for trains going at higher speeds with nuclear fuel, and there might be some attributes that use the slightly faster speed for its calculations, but I've used nuclear fuel trains before and everything worked okay so I'm not sure why suddenly this would be a problem.

Was it confirmed that the problem happens isn't related to that mod and still happens with just RT, and is it on both magnet ramps and regular ramps?

**Comment by Kiplacon** [comment]
Another question, did you only see this happening after your new animation code was put in? I noticed in the other issue post you had a gif showing the spinning carriages and I noticed that the leading locomotive was actually fine and it was only the following carriages that spun. There's a section in the entity_damaged script that transfers some of the information from the leading carriage to its followers. A small difference in that information transfer combined with the higher speeds and new animation code might be what is causing the discrepancy which was unnoticable at slower speeds

**Comment by Kiplacon** [comment]
I've been trying on a test map for a while and I haven't been able to recreate this issue with the spinning carriages or separating trains. Whether slow or fast, trains don't seem to do this, but in the save file from the other post the separation does happen so the issue is legit. But for the time being I can't recreate it to investigate further

**Comment by Xcelled** [comment]
Sorry for the delay, been busy with something else.

I've managed to recreate this on the test map with only RT installed. After a few iterations of the bug not happening, I've identified what seems to be the key cause of the bug: different fuel _types_ in each loco.

I neglected to consider (or even remember) this, but in the save I posted, I would get rid of excess wood by putting it in the train to be burned. Over time, the distribution of the wood became such that the front loco was burning nuclear fuel and the back loco was burning wood. Replicating that into the test setup got the bug to reappear. I'm going to hazard a guess that this only shows up where the train is traveling faster than the second loco's independent max speed.

I've tested the following fuel combinations (using an L-C-L off a magnet ramp):

| Loco 1   | Loco 2   | Bugged?  |
|---|---|---|
| Nuclear | Rocket | No |
| Nuclear | Solid | **Maybe** |
| Nuclear  | Wood  | **Yes**  |
| Nuclear  | None  | **Yes**  |
| Rocket | Nuclear | No |
| Rocket | Solid | **Maybe** |
| Rocket | Wood | **Yes** |
| Rocket | None | **Yes** |
| Solid | Nuclear | No |
| Solid | Rocket | No |
| Solid | Wood | No |
| Solid | None | No |
| Wood | Nuclear | No |
| Wood | Rocket | No |
| Wood | Solid | No |
| Wood | None | No |

Entries with maybe are where the train didn't separate but might have glitched graphically upon landing.

The test map also allows me to confirm that the problem happens for both magnet ramps and regular ramps, though only magnet ramps suffer the graphical spin issue, presumably due to the fact that they change the speed calculations. Magnet ramps also produce the issue across a wider variety of train-fuel combinations, probably for the same reason.

I created the new animation code because I was debugging this issue, but to be safe and ensure that it wasn't my code that caused it, I installed version 0.5.5 of RT (randomly chosen to be "a while ago") and confirmed the issue was still present.

Here's a debug map that reproduces the issue, including both magnet and regular ramps that can be chosen by turning the attached combinator on or off:
[jump test.zip](https://github.com/Kiplacon/Renai-Transportation/files/5580567/jump.test.zip)

**Comment by Kiplacon** [comment]
Wow I never would have thought of that, that's some good detective work 🔍. That would make sense because I'm pretty sure that even with commands you can't set a train's speed to go faster than its allowed max. So locomotives after the first with slower fuels will take slightly longer to reach the ramp and that causes the desync. I'm not certain but maybe even a train with 3 nuclear fuel locomotives can't run as fast as a train with just 1 nuclear fuel locomotive? If that's the case, it seems like the only way to really fix this would be to add a ghost locomotive to the back of a train when it jumps with the fastest possible fuel type. I'll try that to see if it fixes it and for now I'll add it on the mod page as a recognized issue

**Comment by Xcelled** [comment]
An idea I had is to replace the fuel of every loco in the train with the same type as the first loco (and obv restore it when it lands), but I'm not sure if that will take effect immediately or if it's even feasible to do in the mod code without large changes to how the train data is stored.

**Comment by Kiplacon** [comment]
So actually there is a command for trains that use fuels LuaEntity.burner.currently_burning which lets you change what kind of fuel a locomotive is using (and hopefully change its max speed too) which is super convenient cause that's exactly what we're looking to do and I'm already tracking it anyway, would just have to add an "original fuel" property to switch back to. So in theory if the train separation and late landing does come from the difference in fuel types, when a train hits a ramp we would just have to change currently_burning of all locomotives in the train to the fastest possible fuel then when it lands have it switch back. I'm not too concerned if some fuel is gained or lost in that switch. I'll give this a shot over the holidays

**Comment by Kiplacon** [comment]
Should be fixed in 0.6.8. Made a change where when a train hits a ramp, it notes what all locomotives are currently burning before switching them all to the fastest fuel item prototype. Then just before each other locomotive hits the ramp, it switches back.
